pkg/config/agentmanagement.go

```diff
@@ -116,6 +91,78 @@ func fetchConfig(c *Config) ([]byte, error) {
 	return bb, nil
 }
 
+type labelMap map[string]string
+
+type RemoteConfiguration struct {
+	Labels    labelMap `yaml:"labels"`
+	Namespace string   `yaml:"namespace"`
+}
+
+type AgentManagement struct {
+	Enabled         bool             `yaml:"-"` // Derived from enable-features=agent-management
+	Url             string           `yaml:"api_url"`
+	BasicAuth       config.BasicAuth `yaml:"basic_auth"`
+	Protocol        string           `yaml:"protocol"`
+	PollingInterval string           `yaml:"polling_interval"`
+	CacheLocation   string           `yaml:"remote_config_cache_location"`
+
+	RemoteConfiguration RemoteConfiguration `yaml:"remote_configuration"`
+}
+
+// getRemoteConfig gets the remote config specified in the initial config, falling back to a local, cached copy
+// of the remote config if the request to the remote fails. If both fail, an empty config and an
+// error will be returned.
+//
+// It also validates that the response from the API is a semantically correct config by calling config.Validate().
+func getRemoteConfig(expandEnvVars bool, configProvider remoteConfigProvider, log *server.Logger, fs *flag.FlagSet, args []string, configPath string) (*Config, error) {
+	if err := configProvider.AgentManagementConfig().Validate(); err != nil {
+		return nil, fmt.Errorf("invalid initial config: %w", err)
+	}
+	remoteConfigBytes, err := configProvider.FetchRemoteConfig()
+	if err != nil {
+		level.Error(log).Log("msg", "could not fetch from API, falling back to cache", "err", err)
+		return configProvider.GetCachedRemoteConfig(expandEnvVars)
+	}
+	var remoteConfig Config
+
+	err = LoadBytes(remoteConfigBytes, expandEnvVars, &remoteConfig)
+	if err != nil {
+		level.Error(log).Log("msg", "could not load the response from the API, falling back to cache", "err", err)
+		instrumentation.InstrumentInvalidRemoteConfig("invalid_yaml")
+		return configProvider.GetCachedRemoteConfig(expandEnvVars)
+	}
+
+	// this is done in order to validate the config semantically
+	if err = applyIntegrationValuesFromFlagset(fs, args, configPath, &remoteConfig); err != nil {
+		level.Error(log).Log("msg", "could not apply integration version to config", "err", err)
```

Comment:
probably instrument here too with reason "invalid_flags" (arguably this should never happen but who knows)